## Summary:  
This is a project for Code Louisville Data Analysis 2. The data being used comes from the Louisville Metro Open Data website (https://data.louisvilleky.gov). I am using this data to analyze passing restaurant health department scores vs failing restaurant inspection scores in the Louisville Kentucky Metro area.

Hypothesis: My hypothesis is that there are more passing restaurant scores than failing scores. And that there will be less failing scores in the downtown Louisville zip code as there more fine dining restaurants in that area.

## Loading Data

Importing libraries:

In [69]:
import pandas as pd
import numpy as np
import openpyxl

This CSV comes from the Louisville Metro Open Data website. It is info on food establishments that have passing health department scores from the past year. This data only goes back to March 26, 2022:

In [70]:
passing_df = pd.read_csv('Louisville_Metro_KY_-_Restaurant_Inspection_Scores.csv')
passing_df.dropna(subset=['Grade'], inplace=True)
passing_df

EstablishmentID  InspectionID Ins_TypeDesc     EstablishmentName  \
0               31404       1704512      REGULAR  IROQUOIS HIGH SCHOOL   
2               31407       1697970      REGULAR          WESLEY MANOR   
3               31409       1686588      REGULAR   WESTMINSTER TERRACE   
4               31410       1667903      REGULAR         NAZARETH HOME   
7               31410       1706999      REGULAR         NAZARETH HOME   
...               ...           ...          ...                   ...   
6946           116394       1708692      REGULAR            TOGO INDIA   
6948           116395       1701788      REGULAR          SUBWAY #3676   
6950           116395       1703852     FOLLOWUP          SUBWAY #3676   
6955           116396       1709767      REGULAR     CHICK-FIL-A #4679   
6956           116397       1699402      REGULAR    EIFFEL PIZZA (SAC)   

      PlaceName                    Address  Address2        City State    Zip  \
0           NaN           4615 TAYLOR BLVD       NaN  LOUISVILLE    KY  40215   
2           NaN         5012 E MANSLICK RD       NaN  LOUISVILLE    KY  40219   
3           NaN       2116 BUECHEL BANK RD       NaN  LOUISVILLE    KY  40218   
4           NaN            2000 NEWBURG RD       NaN  LOUISVILLE    KY  40205   
7           NaN            2000 NEWBURG RD       NaN  LOUISVILLE    KY  40205   
...         ...                        ...       ...         ...   ...    ...   
6946        NaN  4109 BARDSTOWN RD STE 103       NaN  LOUISVILLE    KY  40218   
6948        NaN         1979 BROWNSBORO RD       NaN  LOUISVILLE    KY  40206   
6950        NaN         1979 BROWNSBORO RD       NaN  LOUISVILLE    KY  40206   
6955        NaN         11801 PLANTSIDE DR       NaN  LOUISVILLE    KY  40299   
6956        NaN            2100 S FLOYD ST       NaN  LOUISVILLE    KY  40208   

                            TypeDescription InspectionDate  score Grade  \
0          SCHOOL CAFETERIA OR FOOD SERVICE   9/28/22 0:00    100     A   
2     FOOD SERVICE (HOSPITAL / NURSING HOME   8/30/22 0:00    100     A   
3     FOOD SERVICE (HOSPITAL / NURSING HOME   7/15/22 0:00     98     A   
4     FOOD SERVICE (HOSPITAL / NURSING HOME   4/26/22 0:00     93     A   
7     FOOD SERVICE (HOSPITAL / NURSING HOME   10/6/22 0:00     95     C   
...                                     ...            ...    ...   ...   
6946                           FOOD SERVICE  10/12/22 0:00     98     A   
6948                           FOOD SERVICE   9/15/22 0:00     94     C   
6950                           FOOD SERVICE   9/22/22 0:00    100     A   
6955                           FOOD SERVICE  10/17/22 0:00    100     A   
6956                           FOOD SERVICE    9/6/22 0:00     97     A   

                NameSearch  Intersection  ObjectId  
0     IROQUOIS HIGH SCHOOL           NaN         1  
2             WESLEY MANOR           NaN         4  
3      WESTMINSTER TERRACE           NaN         6  
4            NAZARETH HOME           NaN         8  
7            NAZARETH HOME           NaN        11  
...                    ...           ...       ...  
6946            TOGO INDIA           NaN     10776  
6948          SUBWAY #3676           NaN     10778  
6950          SUBWAY #3676           NaN     10782  
6955     CHICK-FIL-A #4679           NaN     10789  
6956    EIFFEL PIZZA (SAC)           NaN     10791  

[3954 rows x 17 columns]

This is a separate CSV from the Louisville Metro Open Data website. This is a list of only the businesses that had a failing health department score. The available data only goes up to October 20, 2022:

In [71]:
failed_df = pd.read_csv('Louisville_Metro_KY_-_Inspection_Violations_of_Failed_Restaurants.csv')
failed_df

EstablishmentID          InspectionDate  InspectionID  \
0             100012  2022/04/18 04:00:00+00       1665890   
1              71722  2022/04/12 04:00:00+00       1665082   
2             111936  2022/04/12 04:00:00+00       1664763   
3             111936  2022/04/12 04:00:00+00       1664763   
4             103564  2022/04/15 04:00:00+00       1665549   
..               ...                     ...           ...   
591            62575  2022/07/05 04:00:00+00       1684328   
592            62575  2022/07/05 04:00:00+00       1684328   
593           115197  2022/10/19 04:00:00+00       1710542   
594           115197  2022/10/19 04:00:00+00       1710542   
595           109387  2022/09/30 04:00:00+00       1711519   

                               EstTypeDesc                   premise_name  \
0                             FOOD SERVICE                     SUSHI MARU   
1                             FOOD SERVICE        AFC SUSHI @ KROGER #356   
2                             FOOD SERVICE               HOME RUN BURGERS   
3                             FOOD SERVICE               HOME RUN BURGERS   
4    FOOD SERVICE (HOSPITAL / NURSING HOME         LANDMARK OF LOUISVILLE   
..                                     ...                            ...   
591                           FOOD SERVICE               WHITE CASTLE #30   
592                           FOOD SERVICE               WHITE CASTLE #30   
593  FOOD SERVICE (HOSPITAL / NURSING HOME                 KLONDIKE MANOR   
594  FOOD SERVICE (HOSPITAL / NURSING HOME                 KLONDIKE MANOR   
595       SELF-CONTAINED MOBILE FOOD UNITS  PHILLY'S BEST FROZEN DESSERTS   

     premise_adr1_num premise_adr1_street premise_city premise_state  \
0              4500.0  S HURSTBOURNE PKWY   LOUISVILLE            KY   
1             12501.0      SHELBYVILLE RD   LOUISVILLE            KY   
2              2723.0  S HURSTBOURNE PKWY   LOUISVILLE            KY   
3              2723.0  S HURSTBOURNE PKWY   LOUISVILLE            KY   
4              1155.0        EASTERN PKWY   LOUISVILLE            KY   
..                ...                 ...          ...           ...   
591            2711.0      FERN VALLEY RD   LOUISVILLE            KY   
592            2711.0      FERN VALLEY RD   LOUISVILLE            KY   
593            3802.0         KLONDIKE LN   LOUISVILLE            KY   
594            3802.0         KLONDIKE LN   LOUISVILLE            KY   
595               NaN     MOBILE VAN 7054   LOUISVILLE            KY   

     premise_zip InspectionType  score  InspTypeSpecificViolID  \
0          40299        REGULAR     98                      30   
1          40243        REGULAR     96                      23   
2          40299        REGULAR     89                      18   
3          40299        REGULAR     89                      21   
4          40217        REGULAR     84                      18   
..           ...            ...    ...                     ...   
591        40213        REGULAR     83                      15   
592        40213        REGULAR     83                      21   
593        40218        REGULAR     92                      18   
594        40218        REGULAR     92                      21   
595        40202        REGULAR     95                       8   

                                         ViolationDesc critical_yn  \
0    COMPLIANCE WITH VARIANCE, SPECIALIZED PROCESS ...         Yes   
1                  PROPER COOLING TIME AND TEMPERATURE         Yes   
2        FOOD-CONTACT SURFACES:  CLEANED AND SANITIZED         Yes   
3                     PROPER COLD HOLDING TEMPERATURES         Yes   
4        FOOD-CONTACT SURFACES:  CLEANED AND SANITIZED         Yes   
..                                                 ...         ...   
591                       Food separated and protected          No   
592                   PROPER COLD HOLDING TEMPERATURES         Yes   
593      FOOD-CONTACT SURFACES:  CLEANED AND SA

Due to the restrictions on this data, I downloaded CSV data for March 26, 2022 through October 20, 2022. This is the timeframe of inspections that I will be comparing.

## Cleaning Data

Dropping uneeded and redundant columns from passing_df:

In [72]:
passing_df.drop(['PlaceName', 'Address2', 'Intersection', 'ObjectId', 'NameSearch'], axis='columns', inplace=True)
passing_df

EstablishmentID  InspectionID Ins_TypeDesc     EstablishmentName  \
0               31404       1704512      REGULAR  IROQUOIS HIGH SCHOOL   
2               31407       1697970      REGULAR          WESLEY MANOR   
3               31409       1686588      REGULAR   WESTMINSTER TERRACE   
4               31410       1667903      REGULAR         NAZARETH HOME   
7               31410       1706999      REGULAR         NAZARETH HOME   
...               ...           ...          ...                   ...   
6946           116394       1708692      REGULAR            TOGO INDIA   
6948           116395       1701788      REGULAR          SUBWAY #3676   
6950           116395       1703852     FOLLOWUP          SUBWAY #3676   
6955           116396       1709767      REGULAR     CHICK-FIL-A #4679   
6956           116397       1699402      REGULAR    EIFFEL PIZZA (SAC)   

                        Address        City State    Zip  \
0              4615 TAYLOR BLVD  LOUISVILLE    KY  40215   
2            5012 E MANSLICK RD  LOUISVILLE    KY  40219   
3          2116 BUECHEL BANK RD  LOUISVILLE    KY  40218   
4               2000 NEWBURG RD  LOUISVILLE    KY  40205   
7               2000 NEWBURG RD  LOUISVILLE    KY  40205   
...                         ...         ...   ...    ...   
6946  4109 BARDSTOWN RD STE 103  LOUISVILLE    KY  40218   
6948         1979 BROWNSBORO RD  LOUISVILLE    KY  40206   
6950         1979 BROWNSBORO RD  LOUISVILLE    KY  40206   
6955         11801 PLANTSIDE DR  LOUISVILLE    KY  40299   
6956            2100 S FLOYD ST  LOUISVILLE    KY  40208   

                            TypeDescription InspectionDate  score Grade  
0          SCHOOL CAFETERIA OR FOOD SERVICE   9/28/22 0:00    100     A  
2     FOOD SERVICE (HOSPITAL / NURSING HOME   8/30/22 0:00    100     A  
3     FOOD SERVICE (HOSPITAL / NURSING HOME   7/15/22 0:00     98     A  
4     FOOD SERVICE (HOSPITAL / NURSING HOME   4/26/22 0:00     93     A  
7     FOOD SERVICE (HOSPITAL / NURSING HOME   10/6/22 0:00     95     C  
...                                     ...            ...    ...   ...  
6946                           FOOD SERVICE  10/12/22 0:00     98     A  
6948                           FOOD SERVICE   9/15/22 0:00     94     C  
6950                           FOOD SERVICE   9/22/22 0:00    100     A  
6955                           FOOD SERVICE  10/17/22 0:00    100     A  
6956                           FOOD SERVICE    9/6/22 0:00     97     A  

[3954 rows x 12 columns]

Narrowing results of passing_df to only show locations such as restaurants and not schools or stores:

In [73]:
passing_df.query("TypeDescription == 'FOOD SERVICE'", inplace=True)
passing_df

EstablishmentID  InspectionID Ins_TypeDesc       EstablishmentName  \
9               32571       1666231      REGULAR       LOGAN'S ROADHOUSE   
12              32571       1709812      REGULAR       LOGAN'S ROADHOUSE   
13              32578       1701238      REGULAR        WHITE CASTLE #19   
15              32579       1691802      REGULAR  BONNIE & CLYDE'S PIZZA   
19              32604       1666534      REGULAR               MOBY DICK   
...               ...           ...          ...                     ...   
6946           116394       1708692      REGULAR              TOGO INDIA   
6948           116395       1701788      REGULAR            SUBWAY #3676   
6950           116395       1703852     FOLLOWUP            SUBWAY #3676   
6955           116396       1709767      REGULAR       CHICK-FIL-A #4679   
6956           116397       1699402      REGULAR      EIFFEL PIZZA (SAC)   

                        Address        City State    Zip TypeDescription  \
9                5229 DIXIE HWY  LOUISVILLE    KY  40216    FOOD SERVICE   
12               5229 DIXIE HWY  LOUISVILLE    KY  40216    FOOD SERVICE   
13               7400 DIXIE HWY  LOUISVILLE    KY  40258    FOOD SERVICE   
15               7611 DIXIE HWY  LOUISVILLE    KY  40258    FOOD SERVICE   
19               9150 DIXIE HWY  LOUISVILLE    KY  40272    FOOD SERVICE   
...                         ...         ...   ...    ...             ...   
6946  4109 BARDSTOWN RD STE 103  LOUISVILLE    KY  40218    FOOD SERVICE   
6948         1979 BROWNSBORO RD  LOUISVILLE    KY  40206    FOOD SERVICE   
6950         1979 BROWNSBORO RD  LOUISVILLE    KY  40206    FOOD SERVICE   
6955         11801 PLANTSIDE DR  LOUISVILLE    KY  40299    FOOD SERVICE   
6956            2100 S FLOYD ST  LOUISVILLE    KY  40208    FOOD SERVICE   

     InspectionDate  score Grade  
9      4/19/22 0:00     95     A  
12    10/17/22 0:00     92     A  
13     9/13/22 0:00     96     A  
15      8/4/22 0:00     96     A  
19     4/20/22 0:00    100     A  
...             ...    ...   ...  
6946  10/12/22 0:00     98     A  
6948   9/15/22 0:00     94     C  
6950   9/22/22 0:00    100     A  
6955  10/17/22 0:00    100     A  
6956    9/6/22 0:00     97     A  

[2703 rows x 12 columns]

Some restaurants have received multiple passing scores, but I only want to focus on the most recent inspections. This shows that there are 2,203 establishments coded as Food Service in this passing scores dataset:

In [74]:
passing_food = passing_df.drop_duplicates(subset="EstablishmentID", keep='last')
passing_food


EstablishmentID  InspectionID Ins_TypeDesc       EstablishmentName  \
12              32571       1709812      REGULAR       LOGAN'S ROADHOUSE   
13              32578       1701238      REGULAR        WHITE CASTLE #19   
15              32579       1691802      REGULAR  BONNIE & CLYDE'S PIZZA   
21              32604       1710306      REGULAR               MOBY DICK   
23              32609       1684232      REGULAR       CAPTAIN D'S #3532   
...               ...           ...          ...                     ...   
6893           116481       1707409      REGULAR           SUBWAY #12182   
6946           116394       1708692      REGULAR              TOGO INDIA   
6950           116395       1703852     FOLLOWUP            SUBWAY #3676   
6955           116396       1709767      REGULAR       CHICK-FIL-A #4679   
6956           116397       1699402      REGULAR      EIFFEL PIZZA (SAC)   

                        Address        City State    Zip TypeDescription  \
12               5229 DIXIE HWY  LOUISVILLE    KY  40216    FOOD SERVICE   
13               7400 DIXIE HWY  LOUISVILLE    KY  40258    FOOD SERVICE   
15               7611 DIXIE HWY  LOUISVILLE    KY  40258    FOOD SERVICE   
21               9150 DIXIE HWY  LOUISVILLE    KY  40272    FOOD SERVICE   
23              10101 DIXIE HWY  LOUISVILLE    KY  40272    FOOD SERVICE   
...                         ...         ...   ...    ...             ...   
6893            5469 NEW CUT RD  LOUISVILLE    KY  40214    FOOD SERVICE   
6946  4109 BARDSTOWN RD STE 103  LOUISVILLE    KY  40218    FOOD SERVICE   
6950         1979 BROWNSBORO RD  LOUISVILLE    KY  40206    FOOD SERVICE   
6955         11801 PLANTSIDE DR  LOUISVILLE    KY  40299    FOOD SERVICE   
6956            2100 S FLOYD ST  LOUISVILLE    KY  40208    FOOD SERVICE   

     InspectionDate  score Grade  
12    10/17/22 0:00     92     A  
13     9/13/22 0:00     96     A  
15      8/4/22 0:00     96     A  
21    10/18/22 0:00    100     A  
23      7/5/22 0:00     95     A  
...             ...    ...   ...  
6893   10/7/22 0:00    100     A  
6946  10/12/22 0:00     98     A  
6950   9/22/22 0:00    100     A  
6955  10/17/22 0:00    100     A  
6956    9/6/22 0:00     97     A  

[2203 rows x 12 columns]

Dropping uneeded columns from failed_df:

In [75]:
failed_df.drop(['rpt_area_id','ObjectId'], axis='columns', inplace=True)
failed_df

EstablishmentID          InspectionDate  InspectionID  \
0             100012  2022/04/18 04:00:00+00       1665890   
1              71722  2022/04/12 04:00:00+00       1665082   
2             111936  2022/04/12 04:00:00+00       1664763   
3             111936  2022/04/12 04:00:00+00       1664763   
4             103564  2022/04/15 04:00:00+00       1665549   
..               ...                     ...           ...   
591            62575  2022/07/05 04:00:00+00       1684328   
592            62575  2022/07/05 04:00:00+00       1684328   
593           115197  2022/10/19 04:00:00+00       1710542   
594           115197  2022/10/19 04:00:00+00       1710542   
595           109387  2022/09/30 04:00:00+00       1711519   

                               EstTypeDesc                   premise_name  \
0                             FOOD SERVICE                     SUSHI MARU   
1                             FOOD SERVICE        AFC SUSHI @ KROGER #356   
2                             FOOD SERVICE               HOME RUN BURGERS   
3                             FOOD SERVICE               HOME RUN BURGERS   
4    FOOD SERVICE (HOSPITAL / NURSING HOME         LANDMARK OF LOUISVILLE   
..                                     ...                            ...   
591                           FOOD SERVICE               WHITE CASTLE #30   
592                           FOOD SERVICE               WHITE CASTLE #30   
593  FOOD SERVICE (HOSPITAL / NURSING HOME                 KLONDIKE MANOR   
594  FOOD SERVICE (HOSPITAL / NURSING HOME                 KLONDIKE MANOR   
595       SELF-CONTAINED MOBILE FOOD UNITS  PHILLY'S BEST FROZEN DESSERTS   

     premise_adr1_num premise_adr1_street premise_city premise_state  \
0              4500.0  S HURSTBOURNE PKWY   LOUISVILLE            KY   
1             12501.0      SHELBYVILLE RD   LOUISVILLE            KY   
2              2723.0  S HURSTBOURNE PKWY   LOUISVILLE            KY   
3              2723.0  S HURSTBOURNE PKWY   LOUISVILLE            KY   
4              1155.0        EASTERN PKWY   LOUISVILLE            KY   
..                ...                 ...          ...           ...   
591            2711.0      FERN VALLEY RD   LOUISVILLE            KY   
592            2711.0      FERN VALLEY RD   LOUISVILLE            KY   
593            3802.0         KLONDIKE LN   LOUISVILLE            KY   
594            3802.0         KLONDIKE LN   LOUISVILLE            KY   
595               NaN     MOBILE VAN 7054   LOUISVILLE            KY   

     premise_zip InspectionType  score  InspTypeSpecificViolID  \
0          40299        REGULAR     98                      30   
1          40243        REGULAR     96                      23   
2          40299        REGULAR     89                      18   
3          40299        REGULAR     89                      21   
4          40217        REGULAR     84                      18   
..           ...            ...    ...                     ...   
591        40213        REGULAR     83                      15   
592        40213        REGULAR     83                      21   
593        40218        REGULAR     92                      18   
594        40218        REGULAR     92                      21   
595        40202        REGULAR     95                       8   

                                         ViolationDesc critical_yn  \
0    COMPLIANCE WITH VARIANCE, SPECIALIZED PROCESS ...         Yes   
1                  PROPER COOLING TIME AND TEMPERATURE         Yes   
2        FOOD-CONTACT SURFACES:  CLEANED AND SANITIZED         Yes   
3                     PROPER COLD HOLDING TEMPERATURES         Yes   
4        FOOD-CONTACT SURFACES:  CLEANED AND SANITIZED         Yes   
..                                                 ...         ...   
591                       Food separated and protected          No   
592                   PROPER COLD HOLDING TEMPERATURES         Yes   
593      FOOD-CONTACT SURFACES:  CLEANED AND SA

Narrowing failed_df results to show only restaurants:

In [76]:
failed_df.query("EstTypeDesc == 'FOOD SERVICE'", inplace=True)
failed_df

EstablishmentID          InspectionDate  InspectionID   EstTypeDesc  \
0             100012  2022/04/18 04:00:00+00       1665890  FOOD SERVICE   
1              71722  2022/04/12 04:00:00+00       1665082  FOOD SERVICE   
2             111936  2022/04/12 04:00:00+00       1664763  FOOD SERVICE   
3             111936  2022/04/12 04:00:00+00       1664763  FOOD SERVICE   
12            111186  2022/03/28 04:00:00+00       1661327  FOOD SERVICE   
..               ...                     ...           ...           ...   
588            62575  2022/07/05 04:00:00+00       1684328  FOOD SERVICE   
589            62575  2022/07/05 04:00:00+00       1684328  FOOD SERVICE   
590            62575  2022/07/05 04:00:00+00       1684328  FOOD SERVICE   
591            62575  2022/07/05 04:00:00+00       1684328  FOOD SERVICE   
592            62575  2022/07/05 04:00:00+00       1684328  FOOD SERVICE   

                premise_name  premise_adr1_num premise_adr1_street  \
0                 SUSHI MARU            4500.0  S HURSTBOURNE PKWY   
1    AFC SUSHI @ KROGER #356           12501.0      SHELBYVILLE RD   
2           HOME RUN BURGERS            2723.0  S HURSTBOURNE PKWY   
3           HOME RUN BURGERS            2723.0  S HURSTBOURNE PKWY   
12            ZAM ZAM COFFEE            1808.0          BERRY BLVD   
..                       ...               ...                 ...   
588         WHITE CASTLE #30            2711.0      FERN VALLEY RD   
589         WHITE CASTLE #30            2711.0      FERN VALLEY RD   
590         WHITE CASTLE #30            2711.0      FERN VALLEY RD   
591         WHITE CASTLE #30            2711.0      FERN VALLEY RD   
592         WHITE CASTLE #30            2711.0      FERN VALLEY RD   

    premise_city premise_state  premise_zip InspectionType  score  \
0     LOUISVILLE            KY        40299        REGULAR     98   
1     LOUISVILLE            KY        40243        REGULAR     96   
2     LOUISVILLE            KY        40299        REGULAR     89   
3     LOUISVILLE            KY        40299        REGULAR     89   
12    LOUISVILLE            KY        40215        REGULAR     84   
..           ...           ...          ...            ...    ...   
588   LOUISVILLE            KY        40213        REGULAR     83   
589   LOUISVILLE            KY        40213        REGULAR     83   
590   LOUISVILLE            KY        40213        REGULAR     83   
591   LOUISVILLE            KY        40213        REGULAR     83   
592   LOUISVILLE            KY        40213        REGULAR     83   

     InspTypeSpecificViolID  \
0                        30   
1                        23   
2                        18   
3                        21   
12                        2   
..                      ...   
588                      10   
589                      18   
590                      50   
591                      15   
592                      21   

                                         ViolationDesc critical_yn  \
0    COMPLIANCE WITH VARIANCE, SPECIALIZED PROCESS ...         Yes   
1                  PROPER COOLING TIME AND TEMPERATURE         Yes   
2        FOOD-CONTACT SURFACES:  CLEANED AND SANITIZED         Yes   
3                     PROPER COLD HOLDING TEMPERATURES         Yes   
12                   Certified food protection manager          No   
..                                                 ...         ...   
588  Adequate hand washing facilities supplied and ...          No   
589      FOOD-CONTACT SURFACES:  CLEANED AND SANITIZED         Yes   
590                     Nonfood-contact surfaces clean          No   
591                       Food separated and protected          No   
592                   PROPER COLD HOLDING TEMPERATURES         Yes   

                                    Insp_Viol_Comments  
0    The permit holder is not complying with the HA...  
1    TCS food (brown rice) requiring refrigeration ...  
2     Food-contact surfaces not clean

Each failure type resulted in a row. For example, one business inspection may be 4 rows of data. These rows restated the score multiple times, throwing off my data. To clean this, I am dropping all duplicates. The result shows that there are 132 establishments coded as Food Service in this failed dataset:

In [77]:
failed_food = failed_df.drop_duplicates(subset="InspectionID", keep='last')
failed_food

EstablishmentID          InspectionDate  InspectionID   EstTypeDesc  \
0             100012  2022/04/18 04:00:00+00       1665890  FOOD SERVICE   
1              71722  2022/04/12 04:00:00+00       1665082  FOOD SERVICE   
3             111936  2022/04/12 04:00:00+00       1664763  FOOD SERVICE   
21            111186  2022/03/28 04:00:00+00       1661327  FOOD SERVICE   
22             67026  2022/03/29 04:00:00+00       1662257  FOOD SERVICE   
..               ...                     ...           ...           ...   
548           105793  2022/08/31 04:00:00+00       1698392  FOOD SERVICE   
561           113094  2022/09/29 04:00:00+00       1704934  FOOD SERVICE   
584           106862  2022/08/10 04:00:00+00       1693288  FOOD SERVICE   
585           103751  2022/08/29 04:00:00+00       1697340  FOOD SERVICE   
592            62575  2022/07/05 04:00:00+00       1684328  FOOD SERVICE   

                             premise_name  premise_adr1_num  \
0                              SUSHI MARU            4500.0   
1                 AFC SUSHI @ KROGER #356           12501.0   
3                        HOME RUN BURGERS            2723.0   
21                         ZAM ZAM COFFEE            1808.0   
22   PRP MAGNET SCHOOL OUTSIDE CONCESSION            5901.0   
..                                    ...               ...   
548                           YOKI BUFFET            1700.0   
561                      EL TACO RANCHERO            5015.0   
584                     TOWNEPLACE SUITES            6601.0   
585                            D NALLEY'S             970.0   
592                      WHITE CASTLE #30            2711.0   

    premise_adr1_street premise_city premise_state  premise_zip  \
0    S HURSTBOURNE PKWY   LOUISVILLE            KY        40299   
1        SHELBYVILLE RD   LOUISVILLE            KY        40243   
3    S HURSTBOURNE PKWY   LOUISVILLE            KY        40299   
21           BERRY BLVD   LOUISVILLE            KY        40215   
22         GREENWOOD RD   LOUISVILLE            KY        40258   
..                  ...          ...           ...          ...   
548         ALLIANT AVE   LOUISVILLE            KY        40299   
561     POPLAR LEVEL RD   LOUISVILLE            KY        40219   
584   PARAMOUNT PARK DR   LOUISVILLE            KY        40213   
585            S 3RD ST   LOUISVILLE            KY        40203   
592      FERN VALLEY RD   LOUISVILLE            KY        40213   

    InspectionType  score  InspTypeSpecificViolID  \
0          REGULAR     98                      30   
1          REGULAR     96                      23   
3          REGULAR     89                      21   
21         REGULAR     84                      55   
22         REGULAR     93                      13   
..             ...    ...                     ...   
548        REGULAR     88                      24   
561        REGULAR     82                      34   
584        REGULAR     77                       2   
585        REGULAR     92                      52   
592        REGULAR     83                      21   

                                         ViolationDesc critical_yn  \
0    COMPLIANCE WITH VARIANCE, SPECIALIZED PROCESS ...         Yes   
1                  PROPER COOLING TIME AND TEMPERATURE         Yes   
3                     PROPER COLD HOLDING TEMPERATURES         Yes   
21   Garbage/refuse properly disposed; facilities m...          No   
22     FOOD IN GOOD CONDITION, SAFE, AND UNADULTERATED         Yes   
..                                                 ...         ...   
548  TIME AS A PUBLIC HEALTH CONTROL:  PROCEDURES A...         Yes   
561  Proper cooling methods used; adequate equipmen...          No   
584                  Certified food protection manager          No   
585        PLUMBING INSTALLED; PROPER BACKFLOW DEVICES         Yes   
592                   PROPER COLD HOLDING TEMPERATURES         Yes   

                                    Insp_Viol_Commen

## Merging Data

Merging data on EstablishmentID column:

In [78]:
merged = pd.merge(failed_food, passing_food,on='EstablishmentID', how='left')
merged.drop(['premise_city', 'premise_state', 'City', 'State'], axis =1, inplace=True)
merged

EstablishmentID        InspectionDate_x  InspectionID_x   EstTypeDesc  \
0             100012  2022/04/18 04:00:00+00         1665890  FOOD SERVICE   
1              71722  2022/04/12 04:00:00+00         1665082  FOOD SERVICE   
2             111936  2022/04/12 04:00:00+00         1664763  FOOD SERVICE   
3             111186  2022/03/28 04:00:00+00         1661327  FOOD SERVICE   
4              67026  2022/03/29 04:00:00+00         1662257  FOOD SERVICE   
..               ...                     ...             ...           ...   
127           105793  2022/08/31 04:00:00+00         1698392  FOOD SERVICE   
128           113094  2022/09/29 04:00:00+00         1704934  FOOD SERVICE   
129           106862  2022/08/10 04:00:00+00         1693288  FOOD SERVICE   
130           103751  2022/08/29 04:00:00+00         1697340  FOOD SERVICE   
131            62575  2022/07/05 04:00:00+00         1684328  FOOD SERVICE   

                             premise_name  premise_adr1_num  \
0                              SUSHI MARU            4500.0   
1                 AFC SUSHI @ KROGER #356           12501.0   
2                        HOME RUN BURGERS            2723.0   
3                          ZAM ZAM COFFEE            1808.0   
4    PRP MAGNET SCHOOL OUTSIDE CONCESSION            5901.0   
..                                    ...               ...   
127                           YOKI BUFFET            1700.0   
128                      EL TACO RANCHERO            5015.0   
129                     TOWNEPLACE SUITES            6601.0   
130                            D NALLEY'S             970.0   
131                      WHITE CASTLE #30            2711.0   

    premise_adr1_street  premise_zip InspectionType  score_x  ...  \
0    S HURSTBOURNE PKWY        40299        REGULAR       98  ...   
1        SHELBYVILLE RD        40243        REGULAR       96  ...   
2    S HURSTBOURNE PKWY        40299        REGULAR       89  ...   
3            BERRY BLVD        40215        REGULAR       84  ...   
4          GREENWOOD RD        40258        REGULAR       93  ...   
..                  ...          ...            ...      ...  ...   
127         ALLIANT AVE        40299        REGULAR       88  ...   
128     POPLAR LEVEL RD        40219        REGULAR       82  ...   
129   PARAMOUNT PARK DR        40213        REGULAR       77  ...   
130            S 3RD ST        40203        REGULAR       92  ...   
131      FERN VALLEY RD        40213        REGULAR       83  ...   

                                    Insp_Viol_Comments InspectionID_y  \
0    The permit holder is not complying with the HA...        1708176   
1    TCS food (brown rice) requiring refrigeration ...        1666627   
2      TCS food not maintained at 5°C (41°F) or below.        1705726   
3                            Dumpster plug is missing.        1661327   
4                 expired food needs to be thrown out         1698667   
..                                                 ...            ...   
127  Not following written procedures for using tim...        1698392   
128  Completely covering TCS foods while being cool...        1704887   
129                 No certified food manager present.        1693288   
130        Service sink  needed.  Hot water hand sink.        1697340   
131  TCS food not maintained at 5°C (41°F) or below...        1686504   

    Ins_TypeDesc                     EstablishmentName  \
0       FOLLOWUP                            SUSHI MARU   
1       FOLLOWUP               AFC SUSHI @ KROGER #356   
2       FOLLOWUP                      HOME RUN BURGERS   
3        REGULAR                        ZAM ZAM COFFEE   
4        REGULAR  PRP MAGNET SCHOOL OUTSIDE CONCESSION   
..           ...                                   ...   
127      REGULAR                           YOKI BUFFET   
128      REGULAR                      EL TACO RANCHERO   
129      REGULAR                     TOWNEPLACE SUITES   
130      REGULAR         

All of the failed restaurants had at least 1 passing score. We now have a data frame with 132 restaurants that at some point have both received a failed score and a passing score during this time frame. Score_x are the failing scores while score_y are the passing scores.

## Graphing Data

Exporting my merged datataframe via openpyxl and writer. I will then be reading this xlsx file into Tableau to make 3 graphs:

In [79]:
writer = pd.ExcelWriter('merged_scores_data.xlsx')
merged.to_excel(writer)
writer.save()
print


/var/folders/3t/xjsq37zj4bd5b4qmbtlgmg6w0000gp/T/ipykernel_720/1542567994.py:3: FutureWarning: save is not part of the public API, usage can give unexpected results and will be removed in a future version
  writer.save()


<function print>

## Conclusion

After cleaning my code in this Tableau notebook, I was able to find out that there were 2,203 restaurants that had a passing score in this timeframe, while 132 of those restaurants received a failing score. Thus proving there are more passing scores given out than failing scores.

-My average fail/pass Tableau graph shows that a restaurant can still average a decent score (typically in the 80s) while still techinically failing and appearing on the failing dataframe.

-My critcal fail pie chart shows that 118 of the failures were due to critical violations and 14 were non critical.

-My heatmap Tableau graph shows that the 40202 zip code only had 1 failing score in this timeframe. This is comparatively lower than other zip codes. Of which the 40213 had the highest amount of failures at 13. Thus proving my second hypothesis.